File: coding/src/com/algos/topics/li10_tree/bt/Bt_L129_SumRootToLeafNumbers.java
package  com.algos.topics.li10_tree.bt;

import com.algos.topics.models.TreeNode;


public class Bt_L129_SumRootToLeafNumbers {

	public static void main(String[] args) {
		Bt_L129_SumRootToLeafNumbers obj = new Bt_L129_SumRootToLeafNumbers();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		System.out.println(obj.sumNumbers(root));
	}

	/*
	 * Problem: https://leetcode.com/problems/sum-root-to-leaf-numbers/
	 * Companies: Amazon
	 * Idea: Nums can be calculated from top down. Hence, calculate the number so far and pass as arg to childs.
	 * At the leave node full nums will be formed.
	 * Sum the two nums formed by left and right branches as they would have formed different nums.
	 * */
	public int sumNumbers(TreeNode root) {
        return sumNumbers(root, 0);
    }

    private int sumNumbers(TreeNode root, int sum){
        if(root == null)
			return 0;

        if(root.left == null && root.right == null)
			return sum*10+root.val;

        return sumNumbers(root.left, sum*10+root.val) + sumNumbers(root.right, sum*10+root.val);
    }

}

File: dsa/patterns/10_trees/bt/python/bt_l129__sum_root_to_leaf_numbers.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/sum-root-to-leaf-numbers/
# Companies: Amazon
# Idea: Numbers can be calculated top-down. Pass the number so far to children.
# At leaf nodes, the full number is formed.
# Sum numbers formed by left and right branches.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def sum_numbers(root):
    def helper(node, current_sum):
        if not node:
            return 0

        current_sum = current_sum * 10 + node.val

        # Leaf node
        if not node.left and not node.right:
            return current_sum

        return helper(node.left, current_sum) + helper(node.right, current_sum)

    return helper(root, 0)


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(sum_numbers(root))  # Expected: 132 + 1365 + 14 = 1511
